### カリキュラム - 【AIコース】DeepLearning入門4 TensorFlowによるNNの実装2

In [1]:
import tensorflow as tf
import numpy as np

/Users/yasu/.pyenv/versions/anaconda3-4.3.0/lib/python3.6/importlib/_bootstrap.py:205: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


In [2]:
x_train = np.array([[0,0], [0,1], [1,0], [1,1]])
y_train = np.array([[0], [1], [1], [0]])

In [3]:
x = tf.placeholder(tf.float32, [None, 2], name = "x")
t = tf.placeholder(tf.float32, [None, 1], name = "t")

In [4]:
W1 = tf.Variable(tf.truncated_normal([2, 2]), name = "W1")
b1 = tf.Variable(tf.zeros([2]), name = "b1")
h = tf.nn.sigmoid(tf.matmul(x, W1) + b1, name = "h")

In [5]:
# モデルの出力の定義
W2 = tf.Variable(tf.truncated_normal([2, 1]), name = "W2")
b2 = tf.Variable(tf.zeros([1]), name = "b2")
y = tf.nn.sigmoid(tf.matmul(h, W2) + b2, name = "y")

In [6]:
# 交差エントロピー誤差の定義
with tf.name_scope("cross_entropy"):
    cross_entropy = -tf.reduce_sum(t * tf.log(y) + (1 - t) * tf.log(1 - y))
# パラメータの最適化
with tf.name_scope("train"):
    train_step = tf.train.GradientDescentOptimizer(0.1).minimize(cross_entropy)
# 正解率の表示
with tf.name_scope("accuracy"):
    correct_prediction = tf.equal(tf.sign(y - 0.5), tf.sign(t - 0.5))
    accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

In [8]:
sess = tf.Session()
sess.run(tf.global_variables_initializer())

tf.summary.scalar("cross_entropy", cross_entropy)
tf.summary.scalar("accuracy", accuracy)
tf.summary.histogram("weights_h", W1)
tf.summary.histogram("biases_h", b1)
tf.summary.histogram("weights_out", W2)
tf.summary.histogram("biases_out", b2)

writer = tf.summary.FileWriter("XOR", sess.graph)
merged = tf.summary.merge_all()

for epoch in range(6000):
    sess.run(train_step, feed_dict = {
        x:x_train,
        t:y_train
    })
    
    if epoch % 100 == 0:
        acc_val = sess.run(
            accuracy, feed_dict = {
                x:x_train,
                t:y_train
            })
        print("epoc: %d, Accuracy: %f" %(epoch, acc_val))
    
    summary, loss_val, acc_val = sess.run([merged, cross_entropy, accuracy], feed_dict = {
        x:x_train,
        t:y_train
    })
    writer.add_summary(summary, epoch)

classified = correct_prediction.eval(session = sess, feed_dict = {
    x:x_train,
    t:y_train
})
prob = y.eval(session = sess, feed_dict = {
    x:x_train,
})

epoc: 0, Accuracy: 0.500000
epoc: 100, Accuracy: 0.500000
epoc: 200, Accuracy: 0.500000
epoc: 300, Accuracy: 0.500000
epoc: 400, Accuracy: 0.750000
epoc: 500, Accuracy: 0.750000
epoc: 600, Accuracy: 0.750000
epoc: 700, Accuracy: 0.750000
epoc: 800, Accuracy: 0.750000
epoc: 900, Accuracy: 1.000000
epoc: 1000, Accuracy: 1.000000
epoc: 1100, Accuracy: 1.000000
epoc: 1200, Accuracy: 1.000000
epoc: 1300, Accuracy: 1.000000
epoc: 1400, Accuracy: 1.000000
epoc: 1500, Accuracy: 1.000000
epoc: 1600, Accuracy: 1.000000
epoc: 1700, Accuracy: 1.000000
epoc: 1800, Accuracy: 1.000000
epoc: 1900, Accuracy: 1.000000
epoc: 2000, Accuracy: 1.000000
epoc: 2100, Accuracy: 1.000000
epoc: 2200, Accuracy: 1.000000
epoc: 2300, Accuracy: 1.000000
epoc: 2400, Accuracy: 1.000000
epoc: 2500, Accuracy: 1.000000
epoc: 2600, Accuracy: 1.000000
epoc: 2700, Accuracy: 1.000000
epoc: 2800, Accuracy: 1.000000
epoc: 2900, Accuracy: 1.000000
epoc: 3000, Accuracy: 1.000000
epoc: 3100, Accuracy: 1.000000
epoc: 3200, Accuracy

In [9]:
print(classified)

[[ True]
 [ True]
 [ True]
 [ True]]


In [10]:
print("W1: ", sess.run(W1))
print("b1: ", sess.run(b1))
print("W2: ", sess.run(W2))
print("b2: ", sess.run(b2))

W1:  [[ 7.17579937  5.31642771]
 [ 7.19384527  5.31976843]]
b1:  [-3.26672959 -8.14001465]
W2:  [[ 11.8873806 ]
 [-12.69848347]]
b2:  [-5.52412271]


In [11]:
# tf.summary.FileWriter("XOR", sess.graph)

##### tensorboardの表示エラー
下記のコマンドをコマンドプロンプトで入力するしたところ、下記のエラーとなった。

コマンド:  
tensorboard --logdir=/Users/yasu/school/curriculum/2018_027/XOR

エラー:  
ensorflow RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6

解決策として、pythonのヴァージョンを3.6から3.5に下げるか、tensorflowのヴァージョンを1.4.0から1.3.0に下げる。  
今回、tensorflowのバージョンを1.3.0に下げた。

参考サイト:  
https://github.com/tensorflow/tensorflow/issues/14182#issuecomment-346132338

tensorflow 1.3.0のインストール:  
pip3 install tensorflow==1.3.0

これにより、safariでも起動ができた
